More sam revisions for v50

clarencechen opened the pull request:
Pull request to attempt a fix for remaining street and SAM issues
- [x]  SAM x Avenue Short T support (contributed by @jflann)
- [x]  Fix Diag STR x Diag Street (see #535)
- [x]  Fix Diag SAM x Diag Road T (SAM terminating)
- [x]  Fix Diag SAM x Diag OWR T (SAM terminating)
- [x]  Fix Diag Road x SAM T (road terminating)
- [x]  Fix Diag OWR x SAM T (OWR terminating)

jflann committed:
add metarules for SAM x Avenue DxO short Ts

jflann committed:
Update meta-generated SAM code

memo33 committed:
script: recompile RRW metarules

clarencechen committed:
Merge branch 'staging' into street-sam-revisions-v50

clarencechen committed:
First draft to update SAM metarules and resolve to fix DxD int far tiles

clarencechen committed:
Add base street DxD end int far tiles to resolver

Temporarily disable new DxD thru int metarules

clarencechen committed:
Compile RUL2 File

clarencechen committed:
Lint

clarencechen committed:
Merge remote-tracking branch 'NAMTeam/street-sam-revisions-v50' into street-sam-revisions-v50

clarencechen committed:
Add note to add exception for new tiles in the segment orientation checker

clarencechen committed:
Update new DxD thru int metarules to avoid base street thru int far tile

clarencechen commented:
@memo33 @jflann @TarkusSC4 I'd love to know what you think of my first contribution to NAM modding whenever you find the time to comment.

memo33 commented:
Apologies for the radio silence. I'm still speechless and impressed that you've figured all this out on your own.

I haven't had a chance to look at the changes in detail. Though, for the long-T issues with the Segment Orientation Checker, that's something we've recently discussed as the [nwm-diag-phase2a](https://github.com/NAMTeam/Network-Addon-Mod/tree/nwm-diag-phase2a) branch has the same type of errors. The current plan is to introduce new custom flags for these long-T situations (201, 203 for diagonals, 202 for orthogonal). For example for the first reported override rule:
```
0x5E59E200,0,0,0x5F504700,0,0=0x5E59E200,0,0,0x5E59E280,0,0
```
instead of:
```
Onewayroad~(0,0,1,3) & Sam2~(0,1,0,0) | Onewayroad~(1,3,0,0) & Street~(3,0,0,0) | % | Onewayroad~(1,3,0,0) & Sam2~(3,0,0,0)
```
the definition would become:
```
Onewayroad~(0,0,1,3) & Sam2~(0,1,203,0) | Onewayroad~(1,3,0,0) & Street~(203,0,0,0) | % | Onewayroad~(1,3,0,0) & Sam2~(203,0,0,0)
```
i.e. the far tile segment of the long-T would be `Sam2~(203,0,0,0)`, which allows distinguishing it from the segment of the short-T `Sam2~(3,0,0,0)` (if that's still a thing). This way, the output tiles of the rule also have matching flags, which satisfies the Segment Orientation Checker. (Even longer T's consisting of more than 2 tiles might need more flags.)

Regarding the actual implementation of these Street and SAM T intersections, that's probably something @jflann will be able to comment more on, being more familiar with the original issues with them.

Also, I think I might break out the 0x5d500100 vs 0x5d302000 changes into a separate PR to have a less noisy diff here.

jflann commented:
I've not been able to review this very carefully yet as I don't have all the development tools set up at the moment.  I will echo what memo said - it is very impressive that you've managed to figure out how this all works.

Have you been able to verify that your changes have the desired effect in game?  That is, have you managed to compile a version of the NAM controller and tested it?  There are scripts in the repository which are designed to build the controllers for release.  You can run this and then load it in game in place of the installed NAM controller.  Otherwise, we have a tool that can display RUL2 code, you can give it just the new portions of the meta-generated code and verify that it looks correct visually.  If you want to try that, come reach out to one of us on the SC4E Discord server so we can point you in the right direction.

Thanks again for your contribution.

clarencechen committed:
adjust guess for SAM-End Road IIDs

fix flipped textures in sam-thru T ints

Add override continuation rules for sam-thru T ints

clarencechen commented:
> I've not been able to review this very carefully yet as I don't have all the development tools set up at the moment. I will echo what memo said - it is very impressive that you've managed to figure out how this all works.
> 
> Have you been able to verify that your changes have the desired effect in game? That is, have you managed to compile a version of the NAM controller and tested it? There are scripts in the repository which are designed to build the controllers for release. You can run this and then load it in game in place of the installed NAM controller. Otherwise, we have a tool that can display RUL2 code, you can give it just the new portions of the meta-generated code and verify that it looks correct visually. If you want to try that, come reach out to one of us on the SC4E Discord server so we can point you in the right direction.
> 
> Thanks again for your contribution.

Hi, thanks for taking the time to review my contributions the RUL2 code. TBH, I don't think I would have been able to do it without @memo33's metarules DSL. I've just compiled this code into a new NAM controller and tested it in the game, but there are some missing textures as I've had to guess some of the IID assignments for the resolver without access to the SAM WIP texture pack you referenced in your PR. Let's connect on the SC4E Discord for the latest textures and the RUL2 visualizer utility, as I'm also interested in contributing to the NWM diagonal T intersection code to accelerate the release of NAM 50.

clarencechen committed:
Update SAM-End x Road Long T IIDs to align with WIP Textures

clarencechen commented:
<img width="1920" height="1200" alt="Sugarloaf-Apr  24, 501761420640" src="https://github.com/user-attachments/assets/c319776c-2217-44aa-9c6c-2adcf450401f" />
The T-End intersections seem to be working well with the latest NAM controller and NAM.DLL

clarencechen committed:
Add rule to convert base Street short T intersections to SAM long T

clarencechen committed:
Fix override continuation rules for Sam-thru T ints

clarencechen committed:
Test new long-T intersection flags with SAMxSAM intersections

clarencechen committed:
Add new long-T intersection flags to SAMxRoad and SAMxOWR DxD intersections

clarencechen committed:
Remove todo with new long-T flags

memo33 commented:
> Also, I think I might break out the 0x5d500100 vs 0x5d302000 changes into a separate PR to have a less noisy diff here.

I've force-pushed, removing 2 commits in favor of #535, as mentioned above.

clarencechen committed:
Merge branch 'staging' into street-sam-revisions-v50

clarencechen committed:
Merge branch 'staging' into street-sam-revisions-v50

clarencechen committed:
Merge branch 'staging' into street-sam-revisions-v50

memo33 committed:
small stability fixes for SAM DxO T's

memo33 reviewed (approved):
Looks perfect to me. I think we can merge this if the textures and paths are complete.

@clarencechen Is anything missing in that regard?

clarencechen committed:
Merge branch 'staging' into street-sam-revisions-v50
